### ISSUE TITLE
PL_sv_undef loses identity

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11928]
Migrated from [rt.perl.org#109726](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=109726) (status was 'resolved')

Searchable as RT109726$

**Comment by p5pRT** [comment]
### From zefram@fysh.org

#### Created by zefram@fysh.org



Passing the standard undef object to a subroutine as an argument fails
to pass by reference\.  Instead the sub sees a fresh \*and writable\*
undefined scalar&#8203;:

$ perl \-lwe 'BEGIN \{ \*x = \\undef; \} sub mutate\($\) \{ print \\$\_\[0\]\, " second"; $\_\[0\] = 456; \} print \\$x\, " first"; mutate\($x\); print "done";'
SCALAR\(0x8dc6500\) first
SCALAR\(0x8dc9758\) second
done

whereas any other scalar has its identity preserved&#8203;:

$ perl \-lwe 'BEGIN \{ \*x = \\123; \} sub mutate\($\) \{ print \\$\_\[0\]\, " second"; $\_\[0\] = 456; \} print \\$x\, " first"; mutate\($x\); print "done";'  
SCALAR\(0x86739f8\) first
SCALAR\(0x86739f8\) second
Modification of a read\-only value attempted at \-e line 1\.
$ perl \-lwe 'BEGIN \{ \*x = \\\!1; \} sub mutate\($\) \{ print \\$\_\[0\]\, " second"; $\_\[0\] = 456; \} print \\$x\, " first"; mutate\($x\); print "done";'
SCALAR\(0x8d5e510\) first
SCALAR\(0x8d5e510\) second
Modification of a read\-only value attempted at \-e line 1\.

The latter case above is PL\_sv\_no\, which is immortal in the same way as
the misbehaving PL\_sv\_undef\.  Devel&#8203;::Peek&#8203;::Dump\(\) confirms the identity
of these special scalars\.  Attempting mutation without the intermediation
of a subroutine call consistently produces the expected failure&#8203;:

$ perl \-lwe 'BEGIN \{ \*x = \\undef; \} $x = 456; print "done";'               
Modification of a read\-only value attempted at \-e line 1\.
$ perl \-lwe 'BEGIN \{ \*x = \\123; \} $x = 456; print "done";'  
Modification of a read\-only value attempted at \-e line 1\.
$ perl \-lwe 'BEGIN \{ \*x = \\\!1; \} $x = 456; print "done";' 
Modification of a read\-only value attempted at \-e line 1\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.15.7:

Configured by zefram at Sat Jan 21 18:33:38 GMT 2012.

Summary of my perl5 (revision 5 version 15 subversion 7) configuration:
   
  Platform:
    osname=linux, osvers=2.6.32-5-686, archname=i386-linux-thread-multi
    uname='linux vigo.rous.org 2.6.32-5-686 #1 smp thu nov 3 04:23:54 utc 2011 i686 gnulinux '
    config_args='-des -Darchname=i386-linux -Dcccdlflags=-fPIC -Dccdlflags=-rdynamic -Dprefix=/home/zefram/usr/perl/perl_install/perl-5.15.7-i32-f52 -Dman1ext=1 -Dman3ext=3perl -Duselargefiles -Dusethreads -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dusedevel -Uversiononly -Ui_db'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.4.5', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/../lib /usr/lib/../lib /lib /usr/lib /usr/lib64
    libs=-lnsl -lgdbm -ldl -lm -lcrypt -lutil -lpthread -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc-2.11.2.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.11.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/home/zefram/usr/perl/perl_install/perl-5.15.7-i32-f52/lib/5.15.7/i386-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.15.7:
    /home/zefram/usr/perl/perl_install/perl-5.15.7-i32-f52/lib/site_perl/5.15.7/i386-linux-thread-multi
    /home/zefram/usr/perl/perl_install/perl-5.15.7-i32-f52/lib/site_perl/5.15.7
    /home/zefram/usr/perl/perl_install/perl-5.15.7-i32-f52/lib/5.15.7/i386-linux-thread-multi
    /home/zefram/usr/perl/perl_install/perl-5.15.7-i32-f52/lib/5.15.7
    .


Environment for perl 5.15.7:
    HOME=/home/zefram
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/zefram/usr/perl/perl_install/perl-5.15.7-i32-f52/bin:/home/zefram/usr/perl/util:/home/zefram/pub/i686-pc-linux-gnu/bin:/home/zefram/pub/common/bin:/usr/bin:/bin:/usr/local/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, Feb 03\, 2012 at 03&#8203;:57&#8203;:41AM \-0800\, Zefram wrote&#8203;:
> Passing the standard undef object to a subroutine as an argument fails
> to pass by reference\.  Instead the sub sees a fresh \*and writable\*
> undefined scalar&#8203;:
> 
> $ perl \-lwe 'BEGIN \{ \*x = \\undef; \} sub mutate\($\) \{ print \\$\_\[0\]\, " second"; $\_\[0\] = 456; \} print \\$x\, " first"; mutate\($x\); print "done";'
> SCALAR\(0x8dc6500\) first
> SCALAR\(0x8dc9758\) second

The PL\_sv\_undef pointer in $x is copied to @&#8203;\_\. When an array element is
fetched in lvalue context \(e\.g\. \\$\_\[0\]\)\, av\_fetch\(\) converts PL\_sv\_undef
int the slot into a real SV&#8203;:

&nbsp;    if \(AvARRAY\(av\)\[key\] == &PL\_sv\_undef\) \{
&nbsp;    emptyness&#8203;:
&nbsp;    if \(lval\)
&nbsp;        return av\_store\(av\,key\,newSV\(0\)\);
&nbsp;    return NULL;
&nbsp;    \}


\-\- 
Dave's first rule of Opera&#8203;:
If something needs saying\, say it&#8203;: don't warble it\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Dave Mitchell wrote&#8203;:
>fetched in lvalue context \(e\.g\. \\$\_\[0\]\)\, av\_fetch\(\) converts PL\_sv\_undef
>int the slot into a real SV&#8203;:

It's another special\-case use of PL\_sv\_undef to mean something
other than the Perl\-visibile standard undef value\.  Another case for
PL\_sv\_placeholder?

\-zefram

**Comment by p5pRT** [comment]
### From @nwc10

I've been thinking about this for a bit\. I've just remembered an
important part\. Right now\, PL\_sv\_placeholder is in perlvars\.h\, \*not\*
intrpvar\.h&#8203;:

/\* Restricted hashes placeholder value\.
 \* The contents are never used\, only the address\. \*/
PERLVAR\(G\, sv\_placeholder\, SV\)

It's a global\, not per\-interpreter\. If I understand things correctly\, the
address &PL\_sv\_placeholder is never actually \*returned\* anywhere \- it's
used to compare against\. That means that nothing can fiddle with its
reference count\, or accidentally upgrade it\, bless it\, or anything else\.

\(I think that Juerd did a lightning talk on undef abuse that included
demonstrating how to bless PL\_sv\_undef\)

Whilst that can be solved by making it per\-interpreter\, I'm not sure what
interesting side effects will be knock ons from there\. eg

On Mon\, Feb 06\, 2012 at 12&#8203;:59&#8203;:33PM \+0000\, Zefram wrote&#8203;:
> Dave Mitchell wrote&#8203;:
> >fetched in lvalue context \(e\.g\. \\$\_\[0\]\)\, av\_fetch\(\) converts PL\_sv\_undef
> >int the slot into a real SV&#8203;:
> 
> It's another special\-case use of PL\_sv\_undef to mean something
> other than the Perl\-visibile standard undef value\.  Another case for
> PL\_sv\_placeholder?

This would mean that &PL\_sv\_placeholder has the potential to be returned
by APIs that didn't use to\. Whilst this is sort of the same problem as
PL\_sv\_undef propagating further than we might like

a\) &PL\_sv\_undef is well known to XS code that wants to care about it
b\) Accidentally propagating PL\_sv\_placeholder onwards to places where it
&nbsp;   shouldn't be \(eg in the front door of the hash APIs\) will cause
&nbsp;   "interesting" bugs\.

&nbsp;   Historically\, PL\_sv\_placeholder was introduced for 5\.8\.1\. 5\.8\.0 added
&nbsp;   restricted hashes \(gah\) and the placeholders\, but used PL\_sv\_undef for
&nbsp;   the placeholder value\. Which broke existing working XS code which stored
&nbsp;   PL\_sv\_undef in hashes\, mostly as a token value where the key was
&nbsp;   important\.

On Fri\, Dec 30\, 2011 at 01&#8203;:40&#8203;:12PM \-0800\, Father Chrysostomos via RT wrote&#8203;:

> There is also this interesting bit&#8203;:
> 
>         else if \(o\->op\_type \!= OP\_METHOD\_NAMED
>         && cSVOPo\->op\_sv == &PL\_sv\_undef\) \{
>         /\* PL\_sv\_undef is hack \- it's unsafe to store it in the
>            AV that is the pad\, because av\_fetch treats values of
>            PL\_sv\_undef as a "free" AV entry and will merrily
>            replace them with a new SV\, causing pad\_alloc to think
>            that this pad slot is free\. \(When\, clearly\, it is not\)
>         \*/
>         SvOK\_off\(PAD\_SVl\(ix\)\);
>         SvPADTMP\_on\(PAD\_SVl\(ix\)\);
>         SvREADONLY\_on\(PAD\_SVl\(ix\)\);
> 
> I don't believe &PL\_sv\_placeholder is ever accessible to Perl\, so using
> that instead of &PL\_sv\_undef in pad\.c would allow XS modules to provide
> an OP\_CONST that actually returns &PL\_sv\_undef\.

This is still somewhat wooly in my head\, but there are several levels in the
hierarchy of what AVs are used for

0&#8203;:   \!AvREAL\(\) && \!AvREIFY\(\) just the stacks? \[and @&#8203;DB&#8203;::args currently\]
1&#8203;:   \!AvREAL\(\) && AvREIFY\(\)  @&#8203;\_ \[and @&#8203;DB&#8203;::args should be here\]
2&#8203;:   Pads
2\.5&#8203;: other AVs seen only by C and XS code that can have AVs\, HVs etc directly
&nbsp;     in them
3&#8203;:   AVs visible to Perl code \- ie via a reference\, and containing only
&nbsp;     scalars and \*references\* to nested containers


As far as can work out\, the stacks store garbage rather than any sort of
pointer for empty parts\. From this&#8203;:

&nbsp;        else \{
&nbsp;        newmax = key \< 3 ? 3 : key;
&nbsp;        MEM\_WRAP\_CHECK\_1\(newmax\+1\, SV\*\, oom\_array\_extend\);
&nbsp;        Newx\(AvALLOC\(av\)\, newmax\+1\, SV\*\);
&nbsp;        ary = AvALLOC\(av\) \+ 1;
&nbsp;        tmp = newmax;
&nbsp;        AvALLOC\(av\)\[0\] = &PL\_sv\_undef;    /\* For the stacks \*/
&nbsp;        \}
&nbsp;        if \(AvREAL\(av\)\) \{
&nbsp;        while \(tmp\)
&nbsp;            ary\[\-\-tmp\] = &PL\_sv\_undef;
&nbsp;        \}

I still don't understand that part commented "For the stacks"\. Why does
the first element\, but the first element \*only\* need to be a valid pointer?


Anyway\, I suspect that there's low risk of problems elsewhere if the Pads
switched to using PL\_sv\_placeholder for a free slot\, as they are only
supposed to be accessed via our APIs\, which would never actually \*return\*
it\. And Pads are already somewhat insane with their manual reference
counting fakery\.

I'm concerned that there would be odd strange breakages if arrays \(more
generally\) started using PL\_sv\_placeholder because it would start escaping\.

But I suspect the only way to get a handle on \*that\* would be to tweak the
core and then smoke CPAN against it\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Feb 09 06&#8203;:13&#8203;:44 2012\, nicholas wrote&#8203;:
> 0&#8203;:   \!AvREAL\(\) && \!AvREIFY\(\) just the stacks? \[and @&#8203;DB&#8203;::args
> currently\]
> 1&#8203;:   \!AvREAL\(\) && AvREIFY\(\)  @&#8203;\_ \[and @&#8203;DB&#8203;::args should be here\]

@&#8203;DB&#8203;::args \*does\* have AvREIFY set\.

> Anyway\, I suspect that there's low risk of problems elsewhere if the
> Pads
> switched to using PL\_sv\_placeholder for a free slot\, as they are only
> supposed to be accessed via our APIs\, which would never actually
> \*return\*
> it\. And Pads are already somewhat insane with their manual reference
> counting fakery\.

Thread cloning would have to know about it\, too\, as @&#8203;DB&#8203;::args can end up
containing pads\.


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

I have fixed this bug in ce0d59fdd and added tests in 97568419b8\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

I have fixed this bug in ce0d59fdd and added tests in 97568419b8\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
